keps/sig-network/4858-ip-cidr-validation/README.md

```diff
@@ -0,0 +1,763 @@
+  dual-stack transition (by allowing an IPv4-only program to be
+  quickly converted to a syntactically-IPv6-only program that was
+  nonetheless semantically dual-stack capable). But they were never
+  widely used, and at this point mostly exist to cause confusion. This
+  confusion could, again, allow for smuggling IP addresses past
+  validation checks.
+
+  **The updated validation will reject IPv4-mapped IPv6 addresses, and
+  CIDR values that contain such IPs.**
+
+- They accept two different kinds of "CIDR values":
+
+    - subnets/masks (e.g., "`192.168.1.0/24`", meaning "all IP
+      addresses whose first 24 bits match the first 24 bits of
+      `192.168.1.0`, which is to say, `192.168.1.0 - 192.168.1.255`")
+
+    - "interface addresses" (or "ifaddrs"), individual IP addresses
+      that are assigned to a particular network interface and thus
+      associated with the subnet attached to that interface (e.g.,
+      "`192.168.1.5/24`", meaning "the IP address `192.168.1.5`, which
+      is on the same network segment as the rest of `192.168.1.0/24`).
+
+  Although the latter type of CIDR string (where there are bits set in
+  the IP address beyond the "prefix length") is often used by CNI
+  plugins, it is not currently used in any Kubernetes API object.
+  Nonethless, `net.ParseCIDR` and `netip.ParsePrefix` accept both
+  types, and so we currently consider the "ifaddr" form to be valid in
+  fields where we are looking for a subnet (e.g.
+  `node.status.podCIDRs`) or a mask (e.g., a NetworkPolicy `ipBlock`).
+  If a string such as "`192.168.1.5/24`" is accepted by Kubernetes and
+  then passed unmodified to another API, it is not always clear
+  whether that API will end up treating it as the equivalent of
+  `192.168.1.0/24` or `192.168.1.5/32`, so again, this could result in
+  problems.
+
+  **The updated validation will require all CIDR values to have only 0
+  bits beyond the prefix length.**
+
+- Though not a problem with _existing_ validation, it is also
+  important to note that the new `netip.ParseAddr` function accepts
+  addresses with "zone identifiers" attached, such as
+  "`fe80::1234%eth0`", meaning "the link-local address `fe80::1234` on
+  the network attached to `eth0` (as opposed to the same link-local
+  address on some other interface)". While specifying zone identifiers
+  is important in some contexts, it should not be necessary for any
+  existing Kubernetes API objects, and would confuse any code that
+  tries to use `netutils.ParseIPSloppy`, so we need to be careful to
+  not accept them.
+
+### Affected Fields
+
+The fields whose validation will be updated are:
+
+  - in `core`:
+    - `endpoints.subsets[].addresses[].ip`
+    - `endpoints.subsets[].notReadyAddresses[].ip`
+    - `node.spec.podCIDRs[]`
+    - `pod.spec.dnsConfig.nameservers[]`
+    - `pod.spec.hostAliases[].ip`
+    - `pod.spec.hostIP`
+    - `pod.spec.hostIPs[]`
+    - `pod.spec.podIP`
+    - `pod.spec.podIPs[].ip`
```

Comment:
s/pod.spec.podIP and hostIP /pod.status.podIP and hostIP/  also I don't think those are a problem because those are set by the kubelet and I don't remember if we already discussed this, but I think we normalize them, or we can normalize them if they are already not normalized, so we can skip warnings at least on PodIPs and HostIPs

Reply:
They're mutable so we need the stricter validation anyway, but I added a section saying we should make sure kubelet and various controllers normalize all of the IPs/CIDRs they write out.